File: HW_2020_09/web/user_data/ui/page_config_editor.py

def make_dropdown(init_value, choices, callback):
	a = html.A(init_value, **{'class':'dropdown-toggle', 'data-toggle':'dropdown', 
								'aria-haspopup':'true', 'aria-expanded':'false', 
								'id':'page_config_editor_select_id_1'})
	ll = html.DIV(**{'class':'dropdown-menu', 'aria-labelledby':'page_config_editor_select_id_1'})
	dropdown = html.TD(a + ll, **{'class':'dropdown'})
	for c in choices:
		li = html.A(c, **{'class':'dropdown-item small'})
		ll <= li
		def on_click(ev):
			a.text = ev.target.text
			callback(ev.target.text)
		li.bind('click', on_click)
	return dropdown

class page_config_editor_obj_py:
	def __init__(self):
		self.type_name = 'page_config_editor'
		self.elt = html.DIV()
		self.event_listener = None
		self.component = None
		self.del_btn = None
	def use_example_data(self):
		self.elt <= html.DIV(self.type_name, **{'class':'alert alert-info', 'role':'alert'})
	def get_data(self, data_name):
		if self.component is None: return None
		return self.component
	def set_data(self, data_name, data):
		self.component = None
		self.elt.clear()
		if data is None: return
		self.component = data
		component = self.component['rows'][0]
		tb = html.TABLE(**{'class':'small table table-sm table-hover'})
		self.elt <= tb
		thead = html.THEAD(**{'class':'thead-dark'})
		tb <= thead
		tbody = html.TBODY()
		tb <= tbody
		tr = html.TR()
		thead <= tr
		tr <= html.TH()
		tr <= html.TH('页面参数')
		tr <= html.TH('设置值')
		# page name
		a1 = html.A(component[0], **{'contenteditable':'true'})
		tr = html.TR(**{'class':'table-info'})
		tbody <= tr
		tr <= html.TD(html.SPAN('page', **{'class':'badge badge-info'}))
		tr <= html.TD('页面名称')
		tr <= html.TD(a1)
		def onblur1(ev):
			nonlocal a1
			text = a1.text.strip()
			if text != component[0]:
				component[0] = text
				self.del_btn.text = '删除页面 "'+text+'"'
				if self.event_listener is not None:
					self.event_listener('change', 'component')	
		a1.bind('blur', onblur1)
		# login option
		tr = html.TR(**{'class':'table-info'})
		tbody <= tr
		tr <= html.TD(html.SPAN('page', **{'class':'badge badge-info'}))
		tr <= html.TD('何时显示')
		text2 = component[1].split(',')
		a2 = html.A(text2[1] if len(text2) > 1 else '', **{'contenteditable':'true'})
		def on_login_option():
			nonlocal text2, a2
			text2 = ','.join(text2)
			if component[1] != text2:
				component[1] = text2
				if self.event_listener is not None:
					self.event_listener('change', 'component')
		def callback(option):
			nonlocal text2
			if option == '用户组':
				text2 = [option,a2.text.strip()]
			else:
				text2 = [option]
			on_login_option()
		tr <= make_dropdown(text2[0], ['总是', '登录后', '登录前', '网站作者', '用户组'], callback)
		# user group
		tr = html.TR(**{'class':'table-info'})
		tbody <= tr
		tr <= html.TD(html.SPAN('page', **{'class':'badge badge-info'}))
		tr <= html.TD('用户组名')
		tr <= html.TD(a2)
		def onblur2(ev):
			nonlocal text2
			if len(text2) > 1:
				text2[1] = a2.text.strip()
				on_login_option()
		a2.bind('blur', onblur2)
		# delete page
		tr = html.TR(**{'class':'table-info'})
		tbody <= tr
		tr <= html.TD()
		tr <= html.TD()
		td = html.TD()
		tr <= td
		self.del_btn = html.BUTTON('删除页面 "'+component[0]+'"', style={'font-size':'smaller'}, **{'type':'button', 'class':'btn-sm btn-dark'})
		td <= self.del_btn
		def delete_page(ev):
			self.set_data(None, None)
			self.event_listener('change', 'component')
		def confirm_delete_page(ev):
			from browser import self as window
			window.confirm_modal('请确认删除页面: ' + component[0], delete_page)
		self.del_btn.bind('click', confirm_delete_page)

File: HW_2020_09/web/user_data/ui/prev/page_config_editor.js
/*
  elt
  mounted(config)
  dispose()
  event_listener
  set_data(data_name, data)
  get_data_json_text(data_name)
  set_var(var_name, var_)
  get_var(var_name)
*/


function page_config_editor_obj_js() {
  var obj = {
    type_name: 'page_config_editor',
    elt: null,
    event_listener: null,
    component: null,

    use_example_data() {
      $('<div class="alert alert-info" role="alert">'+this.type_name+'</div>').appendTo($(this.elt));
    },
    get_var(var_name) {
      if (this.component == null) return null;
      return JSON.stringify(this.component);
    },
    set_var(var_name, var_) {
      this.component = null;
      $(obj.elt).empty();
      if (var_ == null) return;
      this.component = JSON.parse(var_)
      // built table element
      var tb = $('<table></table>', {'class':'small table table-sm table-hover '}).appendTo($(obj.elt));
      var thead = $('<thead></thead>', {'class':'thead-dark'}).appendTo(tb);
      var tbody = $('<tbody></tbody>').appendTo(tb);
      var tr = $('<tr></tr>').appendTo(thead);
      $('<th></th>').appendTo(tr);
      $('<th>页面参数</th>').appendTo(tr);
      $('<th>设置值</th>').appendTo(tr);
      tr = $('<tr></tr>', {'class':'table-info'}).appendTo(tbody);
      $('<td><span class="badge badge-info">page</span></td>').appendTo(tr);
      $('<td>何时显示</td>').appendTo(tr);
      var dropdown = $('<td></td>', {'class':'dropdown'}).appendTo(tr);
      var a = $('<a></a>', {'text':this.component[1], 'class':'dropdown-toggle', 'data-toggle':'dropdown', 
                  'aria-haspopup':'true', 'aria-expanded':'false', 'id':'page_config_editor_select_id_1'}).appendTo(dropdown);
      var ll = $('<div></div>', {'class':'dropdown-menu', 'aria-labelledby':'page_config_editor_select_id_1'}).appendTo(dropdown);
      choices = ['', 'login', 'logout', 'owner'].forEach(tt => {
        var li = $('<a></a>', {'text':tt, 'class':'dropdown-item small'}).appendTo(ll);
        li.bind('click', ev => {
          a.text(ev.target.text);
          if (this.component[1] != ev.target.text) {
            this.component[1] = ev.target.text;
            this.event_listener('change', 'component');
          }
        });
      });
      tr = $('<tr></tr>', {'class':'table-info'}).appendTo(tbody);
      $('<td></td>').appendTo(tr);
      $('<td></td>').appendTo(tr);
      var td = $('<td></td>').appendTo(tr);
      var del_btn = $('<button type=button class="btn-sm btn-dark" style="font-size:smaller;">删除页面 "'+this.component[0]+'"</button>').appendTo(td);
      del_btn.bind('click', () => {
        confirm_modal('请确认删除页面: ' + this.component[0], () => {
          this.set_var(null, null);
          this.event_listener('change', 'component');
        });
      });
    }

  };
  obj.elt = document.createElement("div");
  // obj.elt.setAttribute("style", "padding:8px")
  return obj;
}

